### PR TITLE
[fix]: [CDS-68709]: Fixing the NAB issues in matrix with Background and other steps

### PR DESCRIPTION
**Pull request by brijeshdhakar**
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/47768)

### CODE REVIEW

**Review by hextechpal — CHANGES_REQUESTED** [review]
> Reviewed 15 of 15 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/47768)* status: 1 change requests. 0 of 1 approvals obtained. 

___
*[`878-ng-common-utilities/src/main/java/io/harness/steps/matrix/StrategyHelper.java` line 46 at r1](https://reviewable.io/reviews/harness/harness-core/47768#-NVJNBvi2-tpElmLSmgR:-NVJNBvj0TlPZyXeK5jq:bzi6u0s) ([raw file](https://github.com/harness/harness-core/blob/fc931d6a2cec8c271c5f6877ba0d568199157ef2/878-ng-common-utilities/src/main/java/io/harness/steps/matrix/StrategyHelper.java#L46)):*
> ```Java
>   @Inject ParallelismStrategyConfigService parallelismStrategyConfigService;
> 
>   public StrategyInfo expandJsonNodes(JsonNode nodeWithStrategy, Optional<Integer> maxExpansionLimit)
> ```

is this method used any where? Can we mark it as deprecated
___
*[`879-pipeline-ci-commons/src/main/java/io/harness/beans/serializer/RunTimeInputHandler.java` line 382 at r1](https://reviewable.io/reviews/harness/harness-core/47768#-NVJNd_bCxaPVuhXrxxo:-NVJNd_bCxaPVuhXrxxp:bstcksr) ([raw file](https://github.com/harness/harness-core/blob/fc931d6a2cec8c271c5f6877ba0d568199157ef2/879-pipeline-ci-commons/src/main/java/io/harness/beans/serializer/RunTimeInputHandler.java#L382)):*
> ```Java
>       return finalMap;
>     }
> ```

@sahilhindwani-harness @archit-harness please review this again
___
*[`979-recaster/src/main/java/io/harness/core/Recaster.java` line 155 at r1](https://reviewable.io/reviews/harness/harness-core/47768#-NVJNjYn5uids-LraYwZ:-NVJNjYn5uids-LraYw_:bgpnsus) ([raw file](https://github.com/harness/harness-core/blob/fc931d6a2cec8c271c5f6877ba0d568199157ef2/979-recaster/src/main/java/io/harness/core/Recaster.java#L155)):*
> ```Java
>         entityMap.put(entry.getKey(), (o instanceof Map) ? fromMap(RecasterMap.cast((Map<String, Object>) o)) : o);
>       } catch (RecasterException ex) {
>         log.info(
> ```

change this to warn
___
*[`clients/pipeline-service/pms-sdk-core/src/main/java/io/harness/plancreator/strategy/StrategyExpressionEvaluator.java` line 55 at r1](https://reviewable.io/reviews/harness/harness-core/47768#-NVJNtzZ55e_8IlbyCRc:-NVJNtzZ55e_8IlbyCRd:bhfrq1e) ([raw file](https://github.com/harness/harness-core/blob/fc931d6a2cec8c271c5f6877ba0d568199157ef2/clients/pipeline-service/pms-sdk-core/src/main/java/io/harness/plancreator/strategy/StrategyExpressionEvaluator.java#L55)):*
> ```Java
>   }
> 
>   public static class StrategyResolveFunctorImpl extends ResolveFunctorImpl {
> ```

Add a todo here to combine this logic with AmbianceExpressionEvaluator
___
*[`clients/pipeline-service/pms-sdk-core/src/main/java/io/harness/plancreator/strategy/StrategyUtils.java` line 284 at r1](https://reviewable.io/reviews/harness/harness-core/47768#-NVJOL51ESwhHXHPbJyS:-NVJOL51ESwhHXHPbJyT:b-bf3pas) ([raw file](https://github.com/harness/harness-core/blob/fc931d6a2cec8c271c5f6877ba0d568199157ef2/clients/pipeline-service/pms-sdk-core/src/main/java/io/harness/plancreator/strategy/StrategyUtils.java#L284)):*
> ```Java
>     EngineExpressionEvaluator evaluator =
>         new StrategyExpressionEvaluator(combinations, currentIteration, totalIteration, itemValue);
>     evaluator.resolve(jsonString, ExpressionMode.RETURN_ORIGINAL_EXPRESSION_IF_UNRESOLVED);
> ```

@raghavharness this needs to happen with FF
___
*[`clients/pipeline-service/pms-sdk-core/src/main/java/io/harness/steps/matrix/MatrixConfigServiceHelper.java` line 124 at r1](https://reviewable.io/reviews/harness/harness-core/47768#-NVJOtxv1ty5O09y7qgr:-NVJOtxv1ty5O09y7qgs:b-xe8etn) ([raw file](https://github.com/harness/harness-core/blob/fc931d6a2cec8c271c5f6877ba0d568199157ef2/clients/pipeline-service/pms-sdk-core/src/main/java/io/harness/steps/matrix/MatrixConfigServiceHelper.java#L124)):*
> ```Java
>       }
>       StrategyUtils.replaceExpressions(o, combinations.get(currentIteration), currentIteration, totalCount, null);
>       JsonNode jsonNode1 = JsonPipelineUtils.asTree(RecastOrchestrationUtils.fromMap((Map<String, Object>) o, cls));
> ```

Add a todo here for the CI team to use the class object directly rather than json node
___
*[`clients/pipeline-service/pms-sdk-core/src/main/java/io/harness/steps/matrix/MatrixConfigServiceHelper.java` line 124 at r1](https://reviewable.io/reviews/harness/harness-core/47768#-NVJP7Xp0CL0hajB9PRB:-NVJP7Xp0CL0hajB9PRC:bd6i63f) ([raw file](https://github.com/harness/harness-core/blob/fc931d6a2cec8c271c5f6877ba0d568199157ef2/clients/pipeline-service/pms-sdk-core/src/main/java/io/harness/steps/matrix/MatrixConfigServiceHelper.java#L124)):*
> ```Java
>       }
>       StrategyUtils.replaceExpressions(o, combinations.get(currentIteration), currentIteration, totalCount, null);
>       JsonNode jsonNode1 = JsonPipelineUtils.asTree(RecastOrchestrationUtils.fromMap((Map<String, Object>) o, cls));
> ```

Add a better name for variable

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/serializer/vm/VmRunTestStepSerializer.java [hunk 1]
```diff
@@ -105,6 +107,8 @@ public VmRunTestStep serialize(RunTestsStepInfo runTestsStepInfo, String identif
     long timeout = TimeoutUtils.getTimeoutInSeconds(parameterFieldTimeout, runTestsStepInfo.getDefaultTimeout());
     Map<String, String> envVars =
         resolveMapParameterV2("envVariables", stepName, identifier, runTestsStepInfo.getEnvVariables(), false);
+    envVars = CIStepInfoUtils.injectAndResolveLoopingVariables(
```
Comment (harshjain12) [inline 332-ci-manager/service/src/main/java/io/harness/ci/execution/serializer/vm/VmRunTestStepSerializer.java:110]: injectAndResolveLoopingVariables should be called behind ff ?

File: 879-pipeline-ci-commons/src/main/java/io/harness/beans/serializer/RunTimeInputHandler.java [hunk 2]
```diff
@@ -365,6 +365,21 @@ public static Map<String, String> resolveMapParameterV2(String fieldName, String
       return Collections.emptyMap();
     }
 
+    Map m = fields.getValue();
```
Comment (harshjain12) [inline 879-pipeline-ci-commons/src/main/java/io/harness/beans/serializer/RunTimeInputHandler.java:368]: put it behind flag

File: 879-pipeline-ci-commons/src/main/java/io/harness/beans/serializer/RunTimeInputHandler.java [hunk 3]
```diff
@@ -365,6 +365,21 @@ public static Map<String, String> resolveMapParameterV2(String fieldName, String
...
+    Map<String, String> finalMap = new HashMap<>();
+    for (Object k : m.keySet()) {
```
Comment (harshjain12) [inline 879-pipeline-ci-commons/src/main/java/io/harness/beans/serializer/RunTimeInputHandler.java:370]: m should not be empty

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/utils/CIStepInfoUtils.java [hunk 4]
```diff
@@ -228,4 +234,44 @@ public static boolean canRunVmStepOnHost(CIStepInfoType ciStepInfoType, StageInf
     String pluginName = ciExecutionConfigService.getContainerlessPluginNameForVM(ciStepInfoType);
     return isNotEmpty(pluginName);
   }
+
+  public static Map<String, String> injectAndResolveLoopingVariables(
```
Comment (harshjain12) [inline 332-ci-manager/service/src/main/java/io/harness/ci/execution/utils/CIStepInfoUtils.java:238]: @brijeshdhakar can you review this method

**Review by brijeshdhakar — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/47768)* status: 16 file(s) to review. 1 change requests. 4. unresolved discussions. 0 of 1 approvals obtained. 

___
*[`979-recaster/src/main/java/io/harness/core/Recaster.java` line 155 at r1](https://reviewable.io/reviews/harness/harness-core/47768#-NVJNjYn5uids-LraYwZ:-NVJbCKX6JeUEBsEDZSX:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/fc931d6a2cec8c271c5f6877ba0d568199157ef2/979-recaster/src/main/java/io/harness/core/Recaster.java#L155)):*
<details><summary><i>Previously, hextechpal (Prashant Pal) wrote…</i></summary><blockquote>

change this to warn
</blockquote></details>

Done.
___
*[`clients/pipeline-service/pms-sdk-core/src/main/java/io/harness/plancreator/strategy/StrategyExpressionEvaluator.java` line 55 at r1](https://reviewable.io/reviews/harness/harness-core/47768#-NVJNtzZ55e_8IlbyCRc:-NVJbDWmAruHppObOQV_:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/fc931d6a2cec8c271c5f6877ba0d568199157ef2/clients/pipeline-service/pms-sdk-core/src/main/java/io/harness/plancreator/strategy/StrategyExpressionEvaluator.java#L55)):*
<details><summary><i>Previously, hextechpal (Prashant Pal) wrote…</i></summary><blockquote>

Add a todo here to combine this logic with AmbianceExpressionEvaluator
</blockquote></details>

Done.
___
*[`clients/pipeline-service/pms-sdk-core/src/main/java/io/harness/steps/matrix/MatrixConfigServiceHelper.java` line 124 at r1](https://reviewable.io/reviews/harness/harness-core/47768#-NVJP7Xp0CL0hajB9PRB:-NVJbGmT6f-NbKUHngL5:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/fc931d6a2cec8c271c5f6877ba0d568199157ef2/clients/pipeline-service/pms-sdk-core/src/main/java/io/harness/steps/matrix/MatrixConfigServiceHelper.java#L124)):*
<details><summary><i>Previously, hextechpal (Prashant Pal) wrote…</i></summary><blockquote>

Add a better name for variable
</blockquote></details>

Done.
___
*[`clients/pipeline-service/pms-sdk-core/src/main/java/io/harness/steps/matrix/MatrixConfigServiceHelper.java` line 124 at r1](https://reviewable.io/reviews/harness/harness-core/47768#-NVJOtxv1ty5O09y7qgr:-NVJbHBs-_Amifyz-lWz:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/fc931d6a2cec8c271c5f6877ba0d568199157ef2/clients/pipeline-service/pms-sdk-core/src/main/java/io/harness/steps/matrix/MatrixConfigServiceHelper.java#L124)):*
<details><summary><i>Previously, hextechpal (Prashant Pal) wrote…</i></summary><blockquote>

Add a todo here for the CI team to use the class object directly rather than json node
</blockquote></details>

Done.

**Comment by brijeshdhakar** [comment]
*[`332-ci-manager/service/src/main/java/io/harness/ci/execution/utils/CIStepInfoUtils.java` line 238 at r3](https://reviewable.io/reviews/harness/harness-core/47768#gh-1192959710:-NVJgMT26_6-WKcOnraY:b-7nnfh) ([raw file](https://github.com/harness/harness-core/blob/0f4c8fa199c6806d1a011359c12cb16f8dc6f616/332-ci-manager/service/src/main/java/io/harness/ci/execution/utils/CIStepInfoUtils.java#L238)):*
<details><summary><i>Previously, harshjain12 (Harsh Jain) wrote…</i></summary><blockquote>

@brijeshdhakar can you review this method
</blockquote></details>

Looks good to me. @sahilhindwani-harness can you also review this once?

**Review by archit-harness — CHANGES_REQUESTED** [review]
> Reviewed 4 of 15 files at r1, 1 of 1 files at r2, 1 of 12 files at r3, 1 of 3 files at r4, 8 of 10 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/47768)* status: 14 file(s) to review. 2 change requests. 7. unresolved discussions. 0 of 1 approvals obtained. 

___
*[`332-ci-manager/service/src/main/java/io/harness/ci/execution/utils/CIStepInfoUtils.java` line 262 at r5](https://reviewable.io/reviews/harness/harness-core/47768#-NVJkLRr5ivMaNviFljy:-NVJkLRs1eihX3cJH4Hb:bz5godl) ([raw file](https://github.com/harness/harness-core/blob/4540d0379f839e7e1a9ab7cf46d7d8b107dc0045/332-ci-manager/service/src/main/java/io/harness/ci/execution/utils/CIStepInfoUtils.java#L262)):*
> ```Java
>           harnessNodeIndex = stepCurrentIteration;
>           harnessNodeTotal = stepTotalIterations;
>         } else if (stageTotalIterations > 1) {
> ```

@brijeshdhakar can you confirm this as i don't see this code anywhere else
___
*[`879-pipeline-ci-commons/src/main/java/io/harness/beans/serializer/RunTimeInputHandler.java` line 382 at r1](https://reviewable.io/reviews/harness/harness-core/47768#-NVJNd_bCxaPVuhXrxxo:-NVJl0d54mMz91IXv4UZ:b1orqho) ([raw file](https://github.com/harness/harness-core/blob/fc931d6a2cec8c271c5f6877ba0d568199157ef2/879-pipeline-ci-commons/src/main/java/io/harness/beans/serializer/RunTimeInputHandler.java#L382)):*
<details><summary><i>Previously, hextechpal (Prashant Pal) wrote…</i></summary><blockquote>

@sahilhindwani-harness @archit-harness please review this again
</blockquote></details>

verified looks good
___
*[`879-pipeline-ci-commons/src/main/java/io/harness/beans/serializer/RunTimeInputHandler.java` line 368 at r3](https://reviewable.io/reviews/harness/harness-core/47768#gh-1192959126:-NVJkc8KAlK5IbhAFUar:b-wvvzrq) ([raw file](https://github.com/harness/harness-core/blob/0f4c8fa199c6806d1a011359c12cb16f8dc6f616/879-pipeline-ci-commons/src/main/java/io/harness/beans/serializer/RunTimeInputHandler.java#L368)):*
<details><summary><i>Previously, harshjain12 (Harsh Jain) wrote…</i></summary><blockquote>

put it behind flag
</blockquote></details>

can you add comments why this map handling there to handle for expression for complete object, this is for us as well
___
*[`879-pipeline-ci-commons/src/main/java/io/harness/beans/serializer/RunTimeInputHandler.java` line 379 at r5](https://reviewable.io/reviews/harness/harness-core/47768#-NVJl44a92qfQ1qCOcAo:-NVJl44b75eFdoG6PGXw:b-mv4y8y) ([raw file](https://github.com/harness/harness-core/blob/4540d0379f839e7e1a9ab7cf46d7d8b107dc0045/879-pipeline-ci-commons/src/main/java/io/harness/beans/serializer/RunTimeInputHandler.java#L379)):*
> ```Java
>     }
> 
>     if (finalMap.size() > 0) {
> ```

add comment here that we are returning as above method would have get all values for complete object being expression
___
*[`979-recaster/src/main/java/io/harness/core/Recaster.java` line 156 at r5](https://reviewable.io/reviews/harness/harness-core/47768#-NVJlFoE96MEWkuuddiM:-NVJlFoE96MEWkuuddiN:b-z4t6vn) ([raw file](https://github.com/harness/harness-core/blob/4540d0379f839e7e1a9ab7cf46d7d8b107dc0045/979-recaster/src/main/java/io/harness/core/Recaster.java#L156)):*
> ```Java
>       } catch (RecasterException ex) {
>         log.warn(
>             "RecasterKey not present in the RecasterMap. So putting the original recasterMap as the value inside EntityMap");
> ```

can you add key name here as well in logs @brijeshdhakar 
___
*[`980-commons/src/main/java/io/harness/expression/EngineExpressionEvaluator.java` line 645 at r5](https://reviewable.io/reviews/harness/harness-core/47768#-NVJlM4o5pRC9tL5TXaW:-NVJlM4o5pRC9tL5TXaX:bgevclb) ([raw file](https://github.com/harness/harness-core/blob/4540d0379f839e7e1a9ab7cf46d7d8b107dc0045/980-commons/src/main/java/io/harness/expression/EngineExpressionEvaluator.java#L645)):*
> ```Java
>           // Using this expressionMode to return the value as json. Needed to control this flow via a feature-flag. This
>           // expression-mode will be removed.
>           if (value instanceof Map || value instanceof List || value instanceof String[]) {
> ```

can you try instanceof Collection -> i verified it handles all collection including these 3 except int[] which we don't want to handle

**Review by archit-harness — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/47768)* status: 14 file(s) to review. 2 change requests. 7. unresolved discussions. 0 of 1 approvals obtained. 

___
*[`980-commons/src/main/java/io/harness/text/resolver/ExpressionResolver.java` line 24 at r5](https://reviewable.io/reviews/harness/harness-core/47768#-NVJmKin5BhiTN78vrJ_:-NVJmKin5BhiTN78vrJa:bwmwypk) ([raw file](https://github.com/harness/harness-core/blob/4540d0379f839e7e1a9ab7cf46d7d8b107dc0045/980-commons/src/main/java/io/harness/text/resolver/ExpressionResolver.java#L24)):*
> ```Java
>     // null value(means unresolved), do not replace with null value, instead keep the original expression as is.
>     if ((getExpressionMode() == ExpressionMode.RETURN_ORIGINAL_EXPRESSION_IF_UNRESOLVED
>             || getExpressionMode() == ExpressionMode.AS_JSON_IF_VALUE_IS_COLLECTION)
> ```

can you add this handling in evaluateExpressionBlock line in engineExpressionEvaluator and evaluatePrefixCombinations and resolveInternal in ExpressionEvaluatorUtils

**Review by archit-harness — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/47768)* status: 17 file(s) to review. 2 change requests. 7. unresolved discussions. 0 of 1 approvals obtained. 

___
*[`980-commons/src/main/java/io/harness/text/resolver/ExpressionResolver.java` line 24 at r5](https://reviewable.io/reviews/harness/harness-core/47768#-NVJmKin5BhiTN78vrJ_:-NVJqzj41eMQqgkvjzMC:b-1k4zrd) ([raw file](https://github.com/harness/harness-core/blob/4540d0379f839e7e1a9ab7cf46d7d8b107dc0045/980-commons/src/main/java/io/harness/text/resolver/ExpressionResolver.java#L24)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

can you add this handling in evaluateExpressionBlock line in engineExpressionEvaluator and evaluatePrefixCombinations and resolveInternal in ExpressionEvaluatorUtils
</blockquote></details>

expression mode wont work as all steps have different mode so it will change for them -> lets pass ff in context map

**Comment by sahilhindwani-harness** [comment]
*[`332-ci-manager/config/ci-manager-config.yml` line 373 at r6](https://reviewable.io/reviews/harness/harness-core/47768#-NVJrS-E1-hiOjGg6JCt:-NVJrS-E1-hiOjGg6JCu:b-m5v4ox) ([raw file](https://github.com/harness/harness-core/blob/cc1bd3c5677114c4db302602156d6477f67e384c/332-ci-manager/config/ci-manager-config.yml#L373)):*
> ```YAML
>         name: github.com/drone-plugins/drone-meltwater-cache@refs/tags/v1.5.6
>       gitCloneConfig:
>         name: github.com/wings-software/drone-git@refs/tags/v1.3.2
> ```

nit: why this change?

**Comment by sahilhindwani-harness** [comment]
*[`879-pipeline-ci-commons/src/main/java/io/harness/beans/serializer/RunTimeInputHandler.java` line 379 at r5](https://reviewable.io/reviews/harness/harness-core/47768#-NVJl44a92qfQ1qCOcAo:-NVJsCAi-LkDxUOPEcXy:b32v57) ([raw file](https://github.com/harness/harness-core/blob/4540d0379f839e7e1a9ab7cf46d7d8b107dc0045/879-pipeline-ci-commons/src/main/java/io/harness/beans/serializer/RunTimeInputHandler.java#L379)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

add comment here that we are returning as above method would have get all values for complete object being expression
</blockquote></details>

+1

**Comment by sahilhindwani-harness** [comment]
*[`879-pipeline-ci-commons/src/main/java/io/harness/beans/serializer/RunTimeInputHandler.java` line 382 at r1](https://reviewable.io/reviews/harness/harness-core/47768#-NVJNd_bCxaPVuhXrxxo:-NVJsHiX7PyUbqpdJBkh:bsrudkf) ([raw file](https://github.com/harness/harness-core/blob/fc931d6a2cec8c271c5f6877ba0d568199157ef2/879-pipeline-ci-commons/src/main/java/io/harness/beans/serializer/RunTimeInputHandler.java#L382)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

verified looks good
</blockquote></details>

looks good to me too.

**Comment by sahilhindwani-harness** [comment]
*[`979-recaster/src/main/java/io/harness/core/Recaster.java` line 156 at r5](https://reviewable.io/reviews/harness/harness-core/47768#-NVJlFoE96MEWkuuddiM:-NVJsOMICkhpJ81LrHNA:bsk18sk) ([raw file](https://github.com/harness/harness-core/blob/4540d0379f839e7e1a9ab7cf46d7d8b107dc0045/979-recaster/src/main/java/io/harness/core/Recaster.java#L156)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

can you add key name here as well in logs @brijeshdhakar 
</blockquote></details>

Can we log the exception too?

**Comment by sahilhindwani-harness** [comment]
*[`pipeline-service/modules/orchestration/src/main/java/io/harness/engine/pms/execution/strategy/plannode/PlanNodeExecutionStrategy.java` line 148 at r6](https://reviewable.io/reviews/harness/harness-core/47768#-NVJsriTCMG9GDgTtBiS:-NVJsriTCMG9GDgTtBiT:bmkb2pb) ([raw file](https://github.com/harness/harness-core/blob/cc1bd3c5677114c4db302602156d6477f67e384c/pipeline-service/modules/orchestration/src/main/java/io/harness/engine/pms/execution/strategy/plannode/PlanNodeExecutionStrategy.java#L148)):*
> ```Java
>     ExpressionMode expressionMode = planNode.getExpressionMode();
>     if (pmsFeatureFlagService.isEnabled(
>             AmbianceUtils.getAccountId(ambiance), FeatureName.CI_DISABLE_RESOURCE_OPTIMIZATION)) {
> ```

For now, can we add a condition on stage type as well?
If that FF is enabled then this code flow would also be called for cd. 

Just wanted to have a smaller blast radius.

**Review by archit-harness — COMMENTED** [review]
> Reviewed 1 of 11 files at r6, 10 of 10 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/47768)* status: 15 file(s) to review. 2 change requests. 9. unresolved discussions. 0 of 1 approvals obtained. 

___
*[`980-commons/src/main/java/io/harness/expression/EngineExpressionEvaluator.java` line 645 at r5](https://reviewable.io/reviews/harness/harness-core/47768#-NVJlM4o5pRC9tL5TXaW:-NVK2tOE3cDYAGBDvHHj:bb7f2mf) ([raw file](https://github.com/harness/harness-core/blob/4540d0379f839e7e1a9ab7cf46d7d8b107dc0045/980-commons/src/main/java/io/harness/expression/EngineExpressionEvaluator.java#L645)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

can you try instanceof Collection -> i verified it handles all collection including these 3 except int[] which we don't want to handle
</blockquote></details>

done
___
*[`pipeline-service/modules/orchestration/src/main/java/io/harness/engine/pms/execution/strategy/plannode/PlanNodeExecutionStrategy.java` line 148 at r6](https://reviewable.io/reviews/harness/harness-core/47768#-NVJsriTCMG9GDgTtBiS:-NVK3Ayx-dSSq-yWfiS7:b-hhlalo) ([raw file](https://github.com/harness/harness-core/blob/cc1bd3c5677114c4db302602156d6477f67e384c/pipeline-service/modules/orchestration/src/main/java/io/harness/engine/pms/execution/strategy/plannode/PlanNodeExecutionStrategy.java#L148)):*
<details><summary><i>Previously, sahilhindwani-harness (Sahil Hindwani) wrote…</i></summary><blockquote>

For now, can we add a condition on stage type as well?
If that FF is enabled then this code flow would also be called for cd. 

Just wanted to have a smaller blast radius. 
</blockquote></details>

FF should be fine i guess as only one account would be impacted and we can disable if we see this
___
*[`980-commons/src/main/java/io/harness/text/resolver/ExpressionResolver.java` line 24 at r5](https://reviewable.io/reviews/harness/harness-core/47768#-NVJmKin5BhiTN78vrJ_:-NVK3F_s-eke425JPhLE:bb7f2mf) ([raw file](https://github.com/harness/harness-core/blob/4540d0379f839e7e1a9ab7cf46d7d8b107dc0045/980-commons/src/main/java/io/harness/text/resolver/ExpressionResolver.java#L24)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

expression mode wont work as all steps have different mode so it will change for them -> lets pass ff in context map
</blockquote></details>

done

**Comment by sahilhindwani-harness** [comment]
*[`pipeline-service/modules/orchestration/src/main/java/io/harness/engine/pms/data/PmsEngineExpressionServiceImpl.java` line 67 at r7](https://reviewable.io/reviews/harness/harness-core/47768#-NVK3lwX-6VmGYk482QE:-NVK3lwX-6VmGYk482QF:b-odgeq) ([raw file](https://github.com/harness/harness-core/blob/5e4ce2ee7c809314cf15b31c38773ca9e1987fda/pipeline-service/modules/orchestration/src/main/java/io/harness/engine/pms/data/PmsEngineExpressionServiceImpl.java#L67)):*
> ```Java
>   public Object resolve(Ambiance ambiance, Object o, ExpressionMode expressionMode, List<String> enabledFeaTUREFLAGS) {
>     Map<String, String> contxtMap = new HashMap<>();
>     contxtMap.put(EngineExpressionEvaluator.ENABLED_FEATURE_FLAGS_KEY, String.join(",", enabledFeaTUREFLAGS));
> ```

Add a null check for just in case

**Review by archit-harness — COMMENTED** [review]
> Reviewed 1 of 12 files at r3, 9 of 11 files at r6, 1 of 1 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/47768)* status: 5 file(s) to review. 2 change requests. 8. unresolved discussions. 0 of 1 approvals obtained.

**Review by archit-harness — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/47768)* status: 5 file(s) to review. 1 change requests. 8. unresolved discussions.

**Review by sahilhindwani-harness — DISMISSED** [review]
> Reviewed 1 of 15 files at r1, 4 of 12 files at r3, 8 of 10 files at r5, 11 of 11 files at r6, 9 of 10 files at r7, 1 of 1 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/47768)* status: 1 change requests. 7. unresolved discussions.

**Comment by harshjain12** [comment]
trigger uts

**Review by harshjain12 — DISMISSED** [review]

**Comment by brijeshdhakar** [comment]
trigger messagecheck

**Comment by archit-harness** [comment]
trigger commonchecks

**Comment by archit-harness** [comment]
trigger commonchecks

**Comment by raghavharness** [comment]
trigger uts

**Comment by raghavharness** [comment]
trigger uts

**Pull request merged**